File: problems/0004-median-of-two-sorted-arrays/solution.cpp
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        /*
        Problem: 0004 - Median of Two Sorted Arrays.

        Given two sorted arrays, return their median in O(log(min(m, n))) time.

        Time Complexity: O(log(min(m, n))).
        Space Complexity: O(1).
        */

        vector<int>& A = nums1;
        vector<int>& B = nums2;
        if (A.size() > B.size()) swap(A, B);  // Always binary search the smaller array.

        int m = A.size();
        int n = B.size();
        int total = m + n;
        int half = total / 2;

        int left = 0, right = m;
        while (left <= right) {
            int pA = (left + right) / 2;  // Partition index for A.
            int pB = half - pA;           // Complement partition index for B.

            // Handle boundaries by using sentinel values.
            int Aleft  = (pA > 0) ? A[pA - 1] : INT_MIN;
            int Aright = (pA < m) ? A[pA]     : INT_MAX;
            int Bleft  = (pB > 0) ? B[pB - 1] : INT_MIN;
            int Bright = (pB < n) ? B[pB]     : INT_MAX;

            // Check if partition is valid.
            if (Aleft <= Bright && Bleft <= Aright) {
                // If total number of elements is odd, return the middle element.
                if (total % 2 == 1) {
                    return min(Aright, Bright);
                }
                // If even, return the average of the two middle values.
                return (max(Aleft, Bleft) + min(Aright, Bright)) / 2.0;
            } else if (Aleft > Bright) {
                // Partition A is too far right — Aleft is too large to be in the left half.
                // Move the search range to the left.
                right = pA - 1;
            } else {
                // Partition A is too far left — Aright is too small to separate left and right halves.
                // Move the search range to the right.
                left = pA + 1;
            }
        }

        throw runtime_error("Input arrays are not sorted properly.");
    }
};

File: problems/0004-median-of-two-sorted-arrays/solution.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        """
        Problem: 0004 - Median of Two Sorted Arrays.

        Given two sorted arrays, find their median in O(log(min(m, n))) time.

        Time Complexity: O(log(min(m, n))).
        Space Complexity: O(1).
        """

        A, B = nums1, nums2
        if len(A) > len(B):
            A, B = B, A  # Always binary search the smaller array.

        m, n = len(A), len(B)
        total = m + n
        half = total // 2

        left, right = 0, m
        while left <= right:
            pA = (left + right) // 2  # Partition index in A.
            pB = half - pA            # Partition index in B (complement).

            # Handle out-of-bound values by using ±infinity.
            Aleft = A[pA - 1] if pA > 0 else float('-inf')
            Aright = A[pA] if pA < m else float('inf')
            Bleft = B[pB - 1] if pB > 0 else float('-inf')
            Bright = B[pB] if pB < n else float('inf')

            # Check if the partition is valid.
            if Aleft <= Bright and Bleft <= Aright:
                # Odd total → return the middle value.
                if total % 2:
                    return min(Aright, Bright)
                # Even total → return average of middle two values.
                return (max(Aleft, Bleft) + min(Aright, Bright)) / 2
            elif Aleft > Bright:
                # Partition A is too far right — shift search window left.
                right = pA - 1
            else:
                # Partition A is too far left — shift search window right.
                left = pA + 1

        raise ValueError("Input arrays are not sorted properly.")

